V8: Add umb-icon directive to backoffice to begin transition from icon fonts to svgs

MMasey opened the pull request:
_This was initially a much larger PR so I have decided to recreate it with far fewer commits and code changes to hopefully make it easier to merge The existing icon system is still fully functional so once this is in, we could gradually go through the backoffice and make the changes one by one, rather than one huge monolith of a PR (which is what this one ended up being #5293). The majority of the changes that are part of this PR are from adding the 605 new individual svg icons._ 😜

### Summary
This PR updates almost all of the icons using the icon font found throughout the backoffice with SVGs. Each icon is now it's own svg file with the following naming convention icon-[name]. I have created an IconController that currently pointed to a folder containing these icons and makes them usable via the an angular directive call umb-icon. This not only improves accessibility, but also make it much easier to customise the icons available in the back office.

For this PR I have only updated the IconPicker, and a few other icons that were using the `<ins>` or `<span>` tag. The rest of the changes from `<i>` to `<umb-icon>` can happen later.

One thing that I think needs to be added is the ability to sanitise the SVG's as the IconController reads the svg file directly in order for the angular directive to make use of it. I've found a package that can do HTML Sanitizing that can have additional tags and attributes added to the list of allowed data so that it can work with SVGs too. I have this running without issue on a seperate project and can confirm that it works great.

The package is https://github.com/mganss/HtmlSanitizer

I suppose the issue with is it that it has a dependency on AngleSharp, and I know we're trying to reduce dependencies in v8. 

### `<umb-icon />` directive

```
// The most basic `umb-icon` directive looks something like this
<umb-icon icon-name="icon-profile" />


// You can pass any other attribute through to it as you would any other directive, such as...
<umb-icon icon-name="icon-profile" class="icon icon-profile large" ng-show="model.show"></umb-icon>
```

All of the individual icon svgs now live under `src\Umbraco.Web.UI.Client\src\assets\icons`. When the frontend build process is run these get copied over to the `src\Umbraco.Web.UI\Umbraco\assets` folder, in the same way the other frontend source files do.

These are 2 API endpoints that can be used to get the SVG markup. The controller for these can be found at `v8\src\Umbraco.Web\Editors\IconController.cs`


### GetIcon
returns a single IconModel
```
e.g. - http://umbraco.v8/umbraco/backoffice/UmbracoApi/Icon/GetIcon?iconName=icon-navigation-right

returns
{
    "Name":"icon-navigation-right",
    "SvgString":"<svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 512 512\"><path d=\"M360.124 255.513L148.535 422.442l.002-333.862z\"/></svg>"
}
```

### GetAllIcons 
returns List<IconModel> (all icons found in the icons folder, used for the icon picker)
```
e.g. - http://umbraco.v8/umbraco/backoffice/UmbracoApi/Icon/GetAllIcons

returns
[
{
    "Name":"icon-navigation-right",
    "SvgString":"<svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 512 512\"><path d=\"M360.124 255.513L148.535 422.442l.002-333.862z\"/></svg>"
},
{
    "Name":"icon-navigation-left",
    "SvgString":"<svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 512 512\"><path d=\"M360.124 255.513L148.535 422.442l.002-333.862z\"/></svg>"
}
,...]
```

The beauty of this approach is that if you want to add more (non-core) icons, you can add them to the  `src\Umbraco.Web.UI\Umbraco\assets` folder and you will have access to them. The means for people creating apps for Umbraco will be able to easily add their own icons. The same goes for those wanting to do other personalisations to the backoffice.

In order to see the difference i highly recommend installing the [Font Changer with Google Web Fonts](https://chrome.google.com/webstore/detail/font-changer-with-google/jgjhhoglgjdklldfgoffdiaceffijeke) Chrome extension and view the icon-font version to get an idea of the problem this solves.

I'd really appreciate some input/feedback whether it's around the coding style, tackling the SVG sanitization (I have a potential solution but don't just want to throw it in there) or just general questions. It would also be sweet to get some help in regards to writing some tests etc. 

I've put a fair amount of detail in the issue #3560 so I would recommend having a read through if you'd like some more information.

@nielslyngsoe I know you we're interested in this.

Many thanks

skttl commented:
This is great, thanks @MMasey !

I think we will need a way to add custom icons, without having to put them in `/umbraco/assets`. That feels wrong to me.

Could be adding icons using a package.manifest.

But that is probably another issue.

MMasey commented:
@skttl I completely agree. The goal of the PR is to get the svg icon system in, but adding some configuration to place custom svgs wherever you want is definitely something I’d like to seem. It would make it so much easier for package creators for example to add their own icons etc

Before something like that is added though I feel like getting the SVG sanitisation set up to remove any script tags from the svg files is important. I have a working solution available, but it relies on the HtmlSanitizer package and I don’t know how to add that to the Umbraco project.

nielslyngsoe commented:
@MMasey would you need some assistance from HQ? regarding the HtmlSanitizer package?

MMasey commented:
@nielslyngsoe Yes please, that would be really helpful. Once it's in there I can get it configured so that you can see how it would be integrated, then we could take it from there?

nul800sebastiaan commented:
After [merging @bjarnef's update](https://github.com/umbraco/Umbraco-CMS/pull/6293) @BatJan reminded me of the existence of this. I've merged the current v8 into it, but I'm having a bit of trouble figuring out how to do this one: https://github.com/umbraco/Umbraco-CMS/blob/v8/dev/src/Umbraco.Web.UI.Client/src/views/components/tree/umb-tree-item.html#L9



The additional `span` is there for screen readers and I think it only works when it's inside of the button? Maybe @BatJan can give some advice here!

Other than that, if we can strip any `<script>` tags from custom icon svgs, then that should be sufficient. Remember that the only people who should be able to add custom files are people who have access to modify files, it's much easier to do something malicious with other files if you already have filesystem access.. so I'm not too worried about that. 

Finally, could you have a look at the search icon, it's positioning is a bit off: 

![image](https://user-images.githubusercontent.com/304656/65835540-c6137f00-e2e7-11e9-8859-4eda9312812c.png)

![image](https://user-images.githubusercontent.com/304656/65835548-dd526c80-e2e7-11e9-8fbd-d89a43f2b8cf.png)

![image](https://user-images.githubusercontent.com/304656/65835555-f9560e00-e2e7-11e9-8077-80647dc2213b.png)

BatJan commented:
@nul800sebastiaan @MMasey My take on the referenced code block is that the icon needs to be setup using the `<umb-icon>` directive instead of being attached directly on the `<button>`. It may require a little styling tweak doing it this way but that's the way we can work around it - If we did not need to localize the "Expand..." text at some point `aria-label` could have been used but it will give of trouble later on. So a minor refactor is what I suggest so we end up having something like

```
<button>
<umb-icon></umb-icon>
<span class="sr-only"></span>
</button>
```

nul800sebastiaan commented:
Thanks @BatJan - I meant to say: that's how I did it now: https://github.com/umbraco/Umbraco-CMS/blob/3316bfd5edcf6d3070ca74d0292a106700b18ed2/src/Umbraco.Web.UI.Client/src/views/components/tree/umb-tree-item.html

I had trouble getting the NVDA to speak the expand message, and I don't know how to trigger expansion with my keyboard.

Ah.. but it's not in a `<button>`! That would help, let me try that.

nul800sebastiaan commented:
Alright, keyboard nav works again, but I'm not hearing `Expand child items for {{node.name}}` - code is here: https://github.com/umbraco/Umbraco-CMS/blob/3e2d06a1dc4a660c3cec64758af3593721c9e8fc/src/Umbraco.Web.UI.Client/src/views/components/tree/umb-tree-item.html

I also added `aria-hidden="true"` to the svg icon since I don't believe a screen reader can do anything with it, right?

BatJan commented:
@nul800sebastiaan The missing screen reader announcement might be due to the context, which I can't picture in my mind right now... What I'm saying I'm not sure why it's not happening but it can be related to the context somehow. Personally I would need to take a closer look into it in order to figure out the issue 😃 

Actually screen readers can make sense of SVG's if they're using a `<title>` element but in this context and for now I say let's leave the aria-hidden="true" on it for now 👍

MMasey commented:
Hey @nul800sebastiaan, I think i remember you mentioning that there may be something else that needs to be done before this can be merged in? Was there someone else who needed to review it?

Cheers

nielslyngsoe commented:
Hi @MMasey  I see that Sebastiaan hasn't answered this, but let me ask you then: Do you think anything is missing for this to be merged?
If you think its good I can bring it to a review, would you mind fixing the merge conflicts?
Then I will make sure the ball will be rolling, it's going to be so great to get this one in. Thanks

MMasey commented:
Hi @nielslyngsoe, I think everything is there that needs to be, although it would be good to get a thorough review on your end to make sure :) 

I'll try and get the merge conflicts fixed this week.

Thanks

MMasey commented:
I've fixed the merge conflicts, but have found that since I started this a number of other improvements have created some odd side effects including.

1. If you change the icon in the document type picker, save you changes then go back to the Content section, the tree doesn't update with the correct icon unless you refresh the page.
2. The order that the icons appear sometimes changes, I'm not 100% sure why but is there a specific order that we thing the icons in the icon picker should be? It would be simple enough to make it alphabetic.

Another thing is that one of the aims of the PR it to eventually make it easier to customise the backoffice further by adding in your own icons easily but just dropping svg files into the correct folder.

At the moment this functionality isn't quite there. You'd see the new icon in the icon picker, but you would not see the icon in the Content Tree. If this is something that we think should be part of this PR, please can you let me know all of the areas that a person should be able to customise their icons.

I'm thinking that for the time being it would only be in the following sections

-  Content Tree
-  Nested Content picker
-  Multinode tree picker
-  Multi URL picker

I also wonder if that would be better at a separate PR, with the one mainly being there to get the core functionality in.

Let me know what you think. Thanks

nielslyngsoe commented:
@MMasey I think your right, let's keep this PR as simple as possible and then create the feature to add your own icons as another PR.

In regards to nr.1, nr.3, and nr. 4. do these issues come as part of this PR? or do they also exist in the current version of the dev-branch? If so, we should not try to fix them as part of this, but just make sure that there are issues created on the matter. It's important to keep the scope of each PR as small as possible so we can get things moving :-)

2. In regards to the order, I don't think there are any specific thoughts behind the order today, we should maybe investigate but I don't feel strongly for how it's done today — some might like to keep the current order just for their habits. But if it's not possible to keep the same order I would say alphabetically is a good start. We could always look into improving the icon-picker experience.

Is there anything else needed to get the last bits in place?

BTW. Thanks for your effort on this task, I'm looking forward to getting this in as it will open up for several great new things to happen.

MMasey commented:
Right, it may have taken a while, but I think we're good. I've fixed issue 3 and 4 and have also ordered the icons alphabetically (issue 2).

I've checked issue 1 and it appears that this is already happening, not as a result of this PR so we should be good to go!

bjarnef commented:
@MMasey maybe `icon-name` should be simplified to just `icon` to keep it consistent with other component, e.g. `umb-file-icon` which use `icon-document` så default:
https://github.com/umbraco/Umbraco-CMS/blob/a83b72ed47d9e800e7e50f6c0ffa7753dc9134e7/src/Umbraco.Web.UI.Client/src/common/directives/components/umbfileicon.directive.js#L33

Other components use `icon` attribute as well:
**umbButton**
https://github.com/umbraco/Umbraco-CMS/blob/v8/contrib/src/Umbraco.Web.UI.Client/src/common/directives/components/buttons/umbbutton.directive.js#L65

**umbNodePreview**
https://github.com/umbraco/Umbraco-CMS/blob/v8/contrib/src/Umbraco.Web.UI.Client/src/common/directives/components/umbnodepreview.directive.js#L79

MMasey commented:
Hey @bjarnef, I agree that refactoring `icon-name` to `icon` makes sense. I have update the PR with the change.

I have also fixed a few bugs where using the self closing version e.g. `<umb-icon name="search"/>` would break aspects of the DOM. I'm guessing this is due to the reason stated in https://github.com/angular/angular.js/issues/1953#issuecomment-13135021. I have since updated those instances and updated the inline documentation for the directive accordingly.

Finally i've updated this PR to point to the contrib branch as it was still pointing to dev.

bjarnef commented:
@MMasay yeah, I noticed a self-coding `umb-checkbox` had a side-effect as well in listview custom layouts which didn't render the remove button for custom layouts after the `umb-checkbox`.

It would also make sense to use `icon` instead of `icon-class` for `umb-checkbox` and `umb-radiobutton` as well, but it need to keep `icon-class` and mark this as obsolete.
https://github.com/umbraco/Umbraco-CMS/pull/8445#event-3576698823

MMasey commented:
@bjarnef that makes sense in regards to the `umb-checkbox` although I feel like that is out of the scope of this ticket. Maybe it should be added as a separate issue?

bjarnef commented:
@MMasey yep, just an additional note to the self-closing directive/component issue.

nul800sebastiaan commented:
@MMasey We're getting really, really close with this one, I have a few questions:

- How did you generate the icons? `icon-wrong` ironically seems to be wrong 😅 the rest looks great!

Actually I had another question and just answered it myself! 😉 

I've done some updates:

- Renamed whitelist to allowlist
- Updated the config setting to `Umbraco.Icons.Path` and made it work
- Replaced hardcoded paths to `/umbraco/assets/icons` with the config path and made sure that if you rename the `/umbraco` directory this will still work
- Some small refactoring (just my personal code style preferences, for readability)
- Fixed up `IconController` as you asked in a TODO - made sure it inherits from `UmbracoAuthorizedApiController` for proper backoffice-only access

Top job so far!! If you can fix the `icon-wrong.svg` then we're good to go I think!! 🔥

MMasey commented:
Hi  @nul800sebastiaan . Managed to get a bit of time looking into the icon-wrong issue and have hit an interesting bug.

Essentially I've had to change some legacy icons that were using <span class="icon icon-wrong"></span> to us the <umb-icon> directive. Weirdly though, the directive returns `undefined` for the icon  value. This is on the tour step close button and the some of the buttons in the template editor view.

Maybe it could be to do with the order that the directive is added via the angular DI? I won't get a huge amount of time to look at it this weekend but i'll see if I can get it sorted.

bjarnef commented:
Instead of svg-string attribute, would it be better to use `ng-transclude` like `umb-checkbox` and `umb-radiobutton`?
https://github.com/umbraco/Umbraco-CMS/blob/v8/contrib/src/Umbraco.Web.UI.Client/src/views/components/forms/umb-checkbox.html#L34

With the current behaviour it seems only one `icon` and `svg-string` attributes are neccessary, if `svg-string` is set and has a value, then `icon` is simple ignored.

```
<umb-icon 
        icon="icon-alert"
        svg-string='<svg height="50" width="50"><circle cx="25" cy="25" r="25" fill="red" /></svg>'>
</umb-icon>
```

Instead it could be one of the following:

```
<umb-icon 
        icon="icon-alert"
</umb-icon>
```

```
<umb-icon>
        <svg height="50" width="50"><circle cx="25" cy="25" r="25" fill="red" /></svg>
</umb-icon>
```

The component view would look something like this:

```
<span aria-hidden="true" ng-transclude>
    <i ng-if="iconName" class="{{iconName}}"></i>
</span>
```

Maybe is could also be useful optionally to set size (width and height) of the icon?

```
<umb-icon 
        icon="icon-alert"
        width="32"
        height="32"
</umb-icon>
```

or 

```
<umb-icon>
        <svg height="32" width="32"><circle cx="25" cy="25" r="25" fill="red" /></svg>
</umb-icon>
```

MMasey commented:
Thanks for the heads up @bjarnef 👍 Definitely makes a lot of sense for the refactor, and good spot on the docs error. My aim is to try and work on it tonight for fingers crossed I can get it sorted. 

I need to figure out why the scope keeps returning `undefined` for the icon value too when used in the tours and template area. When i first created this PR i could drop the `umb-icon` directive anywhere and it would just work so I have a feeling something has changed somewhere else.

MMasey commented:
Right, after realising that i was just tired and was using the wrong attribute, i've fixed me `undefined` issue 😂.

@bjarnef In regards to updating the directive to use a `ng-transclude`, i had a little experiment and for the time being I think it would be best to leave it as it is. I think i had at some point tried to do as you said, but i get issues when it's used in the icon picker. I imagine it's something that could be worked out at some point, but I think at this point it would be best to get this branch merged in, and then improve in it from then as a separate PR.

In regards to 

> but the svg should be this instead `<svg height="50" width="50"><circle cx="25" cy="25" r="25" fill="red"></svg>.`

I'm not sure that is correct. The `circle` element should either be self closing or should have a closing `</circle>` tag which isn't needed here. https://developer.mozilla.org/en-US/docs/Web/SVG/Element/circle. I think it's one of those, although it works, it isn't actually valid markup.

I've cleaned up the `request-icon` attribute, I think that was some old markup i had when I was initially working on this (it's been almost 2 years now so it's hard to remember 😂).

I also agree that it would be cool to be able to set a width and a height, although I don't know if it's actually necessary as it's something that could be handled by CSS instead. It may also be a bit complex at we get a string value from the backend for the string, but have to use `$sce.trustAsHtml(svgString)` in order to render it out, which may make it tricky to work with. Again, maybe it would best done as a separate PR, if it would be useful for people.

bjarnef commented:
@MMasey okay, might have a look at it and see if we can use `ng-transclude`.
It seems a bit nicer because you can both overwrite with a custom `<svg>` or custom font icon.

```
<umb-icon>
        <svg height="32" width="32"><circle cx="25" cy="25" r="25" fill="red"></svg>
</umb-icon>
```

```
<umb-icon>
        <i class="fas fa-check"></i>
</umb-icon>
```

Regarding the svg markup you're right. I read it as `<svg height="50" width="50" /></svg>` and didn't notice the start `circle` element 🙈

nul800sebastiaan commented:
FYI: I've removed the application cache for now as it can be quite memory intensive for strings. I think it's not all that necessary anyway since we don't really call out to the icon controller very much, let's see how it goes.

nul800sebastiaan commented:
I think @bjarnef's suggestions are great and I would love if they can be looked at in the future. Maybe Bjarne would like to take a stab at getting the transclude to work, or maybe @MMasey can start a draft PR with his attempt so we work can be done from that one?

For now, this is an awesome new feature already and we can always improve it!

We also talked about getting it to work so that there's a convention for an `App_Plugins` directory so that people can ship their own icons in plugins, and if the icon is named the same as a core one, it would prefer the `App_Plugins` one (so the whole icon set can be overridden). 
Additionally it would be good to have a predictable directory where user icons can go, so you don't need to drop them into the `~/umbraco/assets/icons` folder. Something like `~/assets/umbraco/icons` or something like that.
Those two options closely relate to how we already do it with user translation files: https://our.umbraco.com/documentation/extending/language-files/#package-language-files

Anyway, that is wishes for the future, I think it would be best to get the transclude working before 8.8.0 is done, the rest is progressive enhancements I think, so those are nice to have.

nul800sebastiaan commented:
Alright, we are in a good state now, let's get this merged and we can work on the improvements later!

Love the work @MMasey and the feedback everyone! Very much looking forward to shipping this in v8.8.0!! 🎉 
Thanks for your patience and perseverance @MMasey and awesome job on the implementation! 🏅 ⭐

nul800sebastiaan commented:
FYI Moved the labels to the related issue that is closed through this PR! 👍

mattbrailsford commented:
Question. Is this going to be considered a breaking change?

There is a lot of CSS changes here, and quite a few have added extra qualifiers to css rules, so things like *[icon-] rules are now much stricter requiring icons to be spans or i elements (which granted most will be, but still, if they have previously been available unqualified, this is now not the case and thus breaking).

nul800sebastiaan commented:
@mattbrailsford Have you tested this at all? I was under the impression that we left all the old icon stuff in place but maybe we can still fix some things up if they didn't need to move in the CSS! An example of what breaks for you would be great here!

mattbrailsford commented:
Hey @nul800sebastiaan 

No I haven't but I've just been reviewing the changes. I think for the most part I "follow the rules" of creating icons using `<i class="icon icon-folder"></i>` but the main issue is that tag wasn't a requirement before, so it's feasible that someone wanting to save on markup that needs a clickable icon might have previously had `<a href="..." class="icon icon-info"></a>`. And yes, this is bad for accessibility, but that's not the issue being resolved here. It's switching to SVG icons.

The selectors are also not consistent as all allow the `<i>` tag, but **some** allow  the `<button>` tag too. So there could be issues for people that are using `<button class="icon icon-info"></button>` too where some icons are available, but others aren't, where again, they previously all were.

If the rule is that icons must be `<i>` tags, then this should be the requirement and if it has to be a breaking change, so be it. But if that is the case, we should enforce that rule in the core codebase too. So buttons should become `<button><i class="icon icon-info"></i></button>`.

nul800sebastiaan commented:
Ah cool, it's really hard to anticipate, how people use these things, but I'm sure we can ease up on the strictness! So you're specifically using `<a href="..." class="icon icon-info"></a>` and `<button class="icon icon-info"></button>`?

mattbrailsford commented:
@nul800sebastiaan I'm not myself, but I'm saying "it's not unreasonable to think someone might be" and if changing the specificity of a selector isn't necessary to achieve the end result of the issue (ie, removing the tag requirement would still work), should we really be doing that here.

nul800sebastiaan commented:
Sure thing! I'm guessing @MMasey chose to make them more specific since we only had one type of usage before in Umbraco itself. I didn't think of those scenarios, good catch!

MMasey commented:
Hey @nul800sebastiaan & @mattbrailsford. The specificity for the legacy `icon` classes was done for the reasons mentioned by Seb, but I completely get where you're coming from Matt. I don't think it would be a major issue to do a small refactor to remove that specification, although it does feel like it's clinging on to something that should be changed anyway.

This PR was only the start of what will essentially multiple PRs to eventually go through and replace all old icon font icons with svgs, although we do need to take into account all package creators etc.

I really need to pull my finger out an get some docs sorts too to showcase how the `umb-icon` directive should be used. Ultimately it will hopefully make package creators livers easier to if they want to add their own customs icons too as eventually they won't need to rely on a whole inaccessible icon font, and just their down, beautifully design svg icons ;p 

I'm happy to go with HQs decision on this, where it's declaring it a breaking change, or adding back more legacy support.

mattbrailsford commented:
I’m all for change, I’m just asking for a level of depreciation and procedure like we get in other areas of the codebase so people have a chance to migrate gracefully. CSS currently gets a free pass yet is often the most noticeable type of regression.

MMasey commented:
Yeah I completely agree! I’d hate for this PR to have a detrimental impact to existing Umbraco users, especially if it is most likely to affect package creators as working on a fix and releasing means multiple hurdles for end users and developers.

I do wonder if the simplest and safest thing to do would be to have another PR that updated the legacy icons so they aren’t restricted?

Then once’s we know it’s stable again, we try and document it somewhere to make sure information is visible around how icons work in the Backoffice?

I don’t suppose anyone had any examples of packages that this would break?

mattbrailsford commented:
So, turns out I did have a few `<span class="icon ..."></span>`'s in Vendr (that'll teach me to copy from core codebase) so good job I spotted this 😉

nielslyngsoe commented on src/Umbraco.Web.UI.Client/src/less/helveticons.less:
```diff
@@ -8,35 +8,51 @@
 
-[class^="icon-"],
-[class*=" icon-"] {
+span[class^="icon-"],
+span[class*=" icon-"] {
+    display: inline-block;
+    width: 1em;
+    height: 1em;
+    *margin-right: .3em;
+
+    svg {
+        width: 100%;
+        height: 100%;
+        fill: currentColor;
+    }
+
+    &.large{
+        width: 32px; 
+        height: 32px;
+    }
+    &.medium{
+        width: 24px;
+        height: 24px;
+    }
+    &.small{
+        width: 14px;
+        height: 14px;
+    }
+}
```
Hi @MMasey 
Can I ask you to describe the intention of this piece?
The reason I'm asking because we are looking into reverting parts of this file, to make it possible to have icons in the same manner as previously. aka. spans and buttons with icon-* class. And by making our icons selectors target everything again, I see that colliding with this. Therefore I would remove this part, but let me know the intention of it, what purpose it serves and then we can figure out how this should existing and whether it should.
Thanks in advance, You can also get in touch via Slack if you like?

MMasey replied:
Hey @nielslyngsoe, we may have to update the `umb-icon` to add an attribute or modifier class so that we can detect when we don't want an icon use use the icon font. Maybe something as simple as adding an `icon--svg` when it uses and svg so we can then do something like this

```
.icon {
    // ... icon font code etc

    .icon--svg { // I'm using .icon rather than & here to add a later of specificity to help overwrite the content property
        &::before {
            content: none;
        }
    }
}
```

nielslyngsoe replied:
Yeah, that would be cool. Currently, I'm hacking it by looking for spans with icon, but this approach would be way better.

So I'm already working on reverting these parts. But I never finished or tested my work. So if you like please give this a look, and if you like to change something I hope you have access or can make PR towards this one: https://github.com/umbraco/Umbraco-CMS/pull/8941

nielslyngsoe replied:
@MMasey could you help me out describing what this part is used for. In my revert PR(https://github.com/umbraco/Umbraco-CMS/pull/8941) I didn't find anything using this part so I have removed it. But that feels wrong. So I hope you could help me out, thanks :-)

nul800sebastiaan replied:
the diff has a hard time loading, "this part" is referring to:

![image](https://user-images.githubusercontent.com/304656/94445639-f5cdbf00-01a7-11eb-91b1-889f1fb02de0.png)

MMasey replied:
Hey @nielslyngsoe, it's used to ensure the svg icons have the correct sizing, as I think when it was icon fonts it was using font sizing.

MMasey replied:
Also, did you see the PR I created for your PR at #8958?

nielslyngsoe replied:
Hi @MMasey Thanks, I actually already figured this out, I took this part into the umb-icon.less, I hope this makes sense to you. :-)

nielslyngsoe replied:
I think @nul800sebastiaan is merging the Revert PR now, but please give it a look to see if you think that can work for now? Thanks

MMasey replied:
@nielslyngsoe I've just had another look over your PR at it seems like it has actually remove the functionality to make the svg icons appear in the icon picker.

The revert has renamed `getLegacyIcons` back to `getIcons` in the `iconhelper.service.js` which makes sense, but the other code changes may actually cause issues for backwards compatibility as they were made to help make the fallback work with the `umb-icon`, specifically the formatting of the . Also the `iconpicker.controller.js` no longer utilises the `getAllIcons` method which is responsible to retrieving the svgs.

I think it needs to be something like this (similar to my other PR).

```
iconHelper.getAllIcons()
    .then(icons => {
        vm.icons = icons;
        vm.loading = false;

        iconHelper.getIcons() // Gets legacy icon fronts and appends them to the icon list, removing any duplicates
            .then(icons => {
                if(icons && icons.length > 0) {
                    let legacyIcons = icons.filter(icon => !vm.icons.find(x => x.name == icon.name));
                    vm.icons = legacyIcons.concat(vm.icons);
                }
            });
        });
```


Also, did the the typo correction for the `icon-addressbook.svg` 😉 

I'm happy to have a call to chat about it if it would be easier that going back and forth via this.

nielslyngsoe reviewed (commented):
As @mattbrailsford has mentioned we need to ensure that we don't break packages with this change. Therefore I started a PR reverting part of helveticons.less
Any insights into how we can do this the best way are very welcome, thanks in advance.
PR: https://github.com/umbraco/Umbraco-CMS/pull/8941
